File: 1002.java
import java.util.*;
 
public class Main {
	
	static int[][] map = new int[1002][1002];
	static int[][] sum = new int[1002][1002];
	
	static int max(int p, int q){
		if (p>q){
			return p;
		}
		return q;
	}
	
	public static void main(String[] args){
		
		Scanner input = new Scanner(System.in);
 
		int l,w,a,b;
		int maxnum = -10000;
		
		l = input.nextInt(); w = input.nextInt();
		for (int i = 1; i <= l; i++){
			for (int j = 1; j <= w; j++){
				map[i][j] = input.nextInt();
				sum[i][j] = map[i][j];
				sum[i][j] = sum[i][j]+sum[i-1][j]+sum[i][j-1]-sum[i-1][j-1];
			}
		}
		a = input.nextInt(); b = input.nextInt();
		for (int i = a; i <= l; i++){
			for (int j = b; j <= w; j++){
				maxnum = max(maxnum, sum[i][j]-sum[i-a][j]-sum[i][j-b]+sum[i-a][j-b]);
			}
		}
		input.close();
		System.out.println(maxnum);
	}
}

File: 1002.cpp
#include <iostream>
int l,w,a,b,m,i,j;
int s[1001][1001];
int max(int p,int q){
	if (p>=q) return p; else return q;
}
int main(){
	using namespace std;
	cin>>l>>w;
	for (i=1; i<=l; i++)
		for (j=1; j<=w; j++){
			cin>>s[i][j];
			s[i][j]+=s[i-1][j]+s[i][j-1]-s[i-1][j-1];
		}
	cin>>a>>b;
	for (i=a; i<=l; i++)
		for (j=b; j<=w; j++)
		m=max(m,s[i][j]+s[i-a][j-b]-s[i][j-b]-s[i-a][j]);
	cout<<m;
	return 0;
}